File: java_solutions/96.unique-binary-search-trees.java
/*
 * @lc app=leetcode id=96 lang=java
 *
 * [96] Unique Binary Search Trees
 */

class Solution {
    public int numTrees(int n) {
        return java.util.stream.IntStream.rangeClosed(1, n).mapToObj(i -> java.math.BigInteger.valueOf(n + i)).reduce(java.math.BigInteger.ONE, java.math.BigInteger::multiply).divide(java.util.stream.IntStream.rangeClosed(1, n).mapToObj(i -> java.math.BigInteger.valueOf(i)).reduce(java.math.BigInteger.ONE, java.math.BigInteger::multiply)).divide(java.math.BigInteger.valueOf(n + 1)).intValue();
    }
}

File: kotlin_solutions/96.unique-binary-search-trees.kt
/*
 * @lc app=leetcode id=96 lang=kotlin
 *
 * [96] Unique Binary Search Trees
 */

class Solution {fun numTrees(n: Int): Int = (object { fun count(min: Int, max: Int): Int = if (min > max) 1 else (min..max).sumOf { i -> count(min, i - 1) * count(i + 1, max) } }).count(1, n)}
